fix(core): self-referencing relations `Not unique table/alias`

monrostar opened the pull request:
# Description

This code bypasses the typeorm bug related to an error in duplicate aliases when linking to a parent table
Related issue: https://github.com/vendure-ecommerce/vendure/issues/2738
# Breaking changes

Does this PR include any breaking changes we should be aware of?
* no

# Checklist

📌 Always:
- [x] I have set a clear title
- [x] My PR is small and contains a single feature
- [x] I have [checked my own PR](## "Fix typo's and remove unused or commented out code")

👍 Most of the time:
- [x] I have added or updated test cases
- [ ] I have updated the README if needed

monrostar commented on packages/core/src/service/helpers/list-query-builder/list-query-builder.ts:
```diff
@@ -696,7 +696,17 @@ export class ListQueryBuilder implements OnApplicationBootstrap {
         ) => {
             const currentMetadataIsTreeType = metadata.treeType;
-            if (!currentParentIsTreeType && !currentMetadataIsTreeType) {
+            let currentMetadataHasOneOrMoreSelfReferencingRelations = false;
+            // Check if the current entity has one or more self-referencing relations
+            // to determine if it is a tree type or has tree relations.
+            for (const relation of currentMetadata.relations) {
+                if (relation.inverseEntityMetadata === currentMetadata) {
+                    currentMetadataHasOneOrMoreSelfReferencingRelations = true;
+                    break;
+                }
+            }
+
+            if (!currentParentIsTreeType && !currentMetadataIsTreeType && !currentMetadataHasOneOrMoreSelfReferencingRelations) {
                 return;
             }
```
This is the most important code. It is enough to determine whether it is necessary to do join manually or leave this job to typeorm

monrostar commented on packages/core/e2e/fixtures/test-plugins/list-query-plugin.ts:
```diff
@@ -144,6 +144,12 @@ export class TestEntity extends VendureEntity implements Translatable, HasCustom
 
     @Column(() => TestEntityCustomFields)
     customFields: TestEntityCustomFields;
+
+    @ManyToOne(() => TestEntity, (type) => type.parent)
+    parent: TestEntity | null;
+
+    @Column('int', { nullable: true })
+    parentId: ID | null;
```
Here I had to update the model a bit to repeat the error and take it into account in the tests

tianyingchun replied:
above fix i think need consider more behavior
if we have `translations` like 
```ts
  @OneToMany(
    () => LensProcessOptionTranslation,
    (translation) => translation.base,
    { eager: true }
  )
  translations: Array<Translation<LensProcessOption>>;
```
the `parent` should be auto load `translations` otherwise it will result `translateEntity` failed

```ts

function translateLeaf(
    object: { [key: string]: any } | undefined,
    property: string,
    languageCode: LanguageCode | [LanguageCode, ...LanguageCode[]],
): any {
    if (object && object[property]) {
        if (Array.isArray(object[property])) {
            return object[property].map((nested2: any) => translateEntity(nested2, languageCode));
        } else if (object[property]) {
            return translateEntity(object[property], languageCode);
        }
    }
}

```
if `parent` existed but it have auto loaded `translations` it will broken here

tianyingchun replied:
please see my latest updated repo https://github.com/tianyingchun/next-issue/tree/vendure_issue_2738

tianyingchun replied:
```
InternalServerError [GraphQLError]: error.entity-has-no-translation-in-language
    at new I18nError (/Users/tianyingchun/Documents/coding/kzfoo/vendure-issue/node_modules/@vendure/core/dist/i18n/i18n-error.js:21:9)
    at new InternalServerError (/Users/tianyingchun/Documents/coding/kzfoo/vendure-issue/node_modules/@vendure/core/dist/common/error/errors.js:15:9)
    at translateEntity (/Users/tianyingchun/Documents/coding/kzfoo/vendure-issue/node_modules/@vendure/core/dist/service/helpers/utils/translate-entity.js:33:15)
    at translateLeaf (/Users/tianyingchun/Documents/coding/kzfoo/vendure-issue/node_modules/@vendure/core/dist/service/helpers/utils/translate-entity.js:104:20)
    at translateDeep (/Users/tianyingchun/Documents/coding/kzfoo/vendure-issue/node_modules/@vendure/core/dist/service/helpers/utils/translate-entity.js:89:21)
    at TranslatorService.translate (/Users/tianyingchun/Documents/coding/kzfoo/vendure-issue/node_modules/@vendure/core/dist/service/helpers/translator/translator.service.js:54:53)
    at file:///Users/tianyingchun/Documents/coding/kzfoo/vendure-issue/packages/plugin-issue/src/services/menu.service.ts:82:76
    at Array.map (<anonymous>)
    at file:///Users/tianyingchun/Documents/coding/kzfoo/vendure-issue/packages/plugin-issue/src/services/menu.service.ts:82:37
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  path: undefined,
  locations: undefined,
  extensions: { code: 'INTERNAL_SERVER_ERROR' },
  variables: { entityName: 'Menu', languageCode: ',,en' },
  code: 'INTERNAL_SERVER_ERROR',
  logLevel: 0
}
```

monrostar replied:
Interesting. I thought I had accounted for translations for all self-related relations. I will look into this and add join for eager links

> above fix i think need consider more behavior if we have `translations` like
> 
> ```ts
>   @OneToMany(
>     () => LensProcessOptionTranslation,
>     (translation) => translation.base,
>     { eager: true }
>   )
>   translations: Array<Translation<LensProcessOption>>;
> ```
> 
> the `parent` should be auto load `translations` otherwise it will result `translateEntity` failed
> 
> ```ts
> function translateLeaf(
>     object: { [key: string]: any } | undefined,
>     property: string,
>     languageCode: LanguageCode | [LanguageCode, ...LanguageCode[]],
> ): any {
>     if (object && object[property]) {
>         if (Array.isArray(object[property])) {
>             return object[property].map((nested2: any) => translateEntity(nested2, languageCode));
>         } else if (object[property]) {
>             return translateEntity(object[property], languageCode);
>         }
>     }
> }
> ```
> 
> if `parent` existed but it have auto loaded `translations` it will broken here

monrostar replied:
seems to be a bug in the `translateDeep` function

monrostar commented:
@michaelbromley 
Added a fix. And refactored the code a bit. 

I'm not sure if we should look for translation in nested objects automatically because that would lead to uncontrolled execution of this method. 

`translateDeep` works fine.

tianyingchun commented:
oh it seems that this change also have some issue here

monrostar commented:
Can you also provide the missing code from `ModuleChannelAware`?
At the moment I can see that there might be an error due to missing ReverseSide for relations, but I guess that can be bypassed too.

> ```ts
> export class Module extends VendureEntity implements ModuleChannelAware {
> ```

tianyingchun commented:
```
/**
 * @description
 * Entities which can be assigned to Module Channels should implement this interface.
 */
export interface ModuleChannelAware {
  channels: ModuleChannel[];
}

```

monrostar commented:
@michaelbromley 
I think the attempt to fix the next bug should be done in a separate PR because there are too many changes here already

In this PR, the first bug has been fixed. I'll open a new PR from https://github.com/vendure-ecommerce/vendure/pull/2744
